### PR TITLE
JAMES-1721 SetMailboxes can't modify system mailboxes

### PR DESCRIPTION
**Pull request by Laura-Royet**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 1]
```diff
@@ -71,10 +73,14 @@ private void handleUpdate(String mailboxId, MailboxUpdateRequest updateRequest,
         try {
             validateMailboxName(updateRequest, mailboxSession);
             Mailbox mailbox = getMailbox(mailboxId, mailboxSession);
+            checkRole(mailbox.getRole());
             validateParent(mailbox, updateRequest, mailboxSession);
 
             updateMailbox(mailbox, updateRequest, mailboxSession);
             responseBuilder.updated(mailboxId);
+
+        } catch (SystemMailboxException e) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: It's not a good exception name, It doesn't describe what is wrong

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 2]
```diff
@@ -71,10 +73,14 @@ private void handleUpdate(String mailboxId, MailboxUpdateRequest updateRequest,
...
+            notUpdated(mailboxId, "invalidArguments", "Cannot rename a system mailbox.", responseBuilder);
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: what if it's not a rename but another change ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 3]
```diff
@@ -1441,4 +1441,39 @@ public void setMailboxesShouldUpdateMailboxNameWhenRenamingAChildMailbox() {
             .body(ARGUMENTS + ".list", hasSize(1))
             .body(ARGUMENTS + ".list[0].name", equalTo("mySecondBox"));
     }
+
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: What happens if I rename a mailbox "any" to a system mailbox ? MailboxAlreadyExist exception ?

Maybe a test about this might be nice, as it is some kind of corner case ....

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 4]
```diff
@@ -71,10 +73,14 @@ private void handleUpdate(String mailboxId, MailboxUpdateRequest updateRequest,
         try {
             validateMailboxName(updateRequest, mailboxSession);
             Mailbox mailbox = getMailbox(mailboxId, mailboxSession);
+            checkRole(mailbox.getRole());
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: I don't like this code snipset much...

I would prefer a plain old else / if....

IMO, if you throw an exception, it means your class encountered a problem it can not solve, and relies on the caller to handle it. Here you throw an exception in a method, that is handled by an other pethod, which means your class knows how to handle the exception.

http://stackoverflow.com/questions/8621762/java-if-vs-try-catch-overhead

The important part for me is :

```
try/catch is for things that go wrong that are outside of your control and not in the normal program flow. 


if statements should be normal flow and ordinary error checking. So, for example, user fails to populate
 a required input field? Use if for that, not try/catch.
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 3 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: In this case, the exception will be a MailboxExistExeption without distinction between a target mailbox which would be a system mailbox and a target mailbox which would not be  a system mailbox.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 4 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: Here we have adopted this solution because we didn't want to multiply the number of lines by adding a lot of 'if'.
Moreover, using exceptions in this context allows us  to be in a functional logic.

Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: I partially agree : a method should be as short as possible. when you have to handle a lot of preconditions, it doesn't make sense to handle them with if statements. On the other side, I agree that we should not abuse exception for local matters.
The right solution is probably to have a precondition method that return a list of errors, then we would do a if on the error list. But the cost of doing this here is high.
Finally, if I quote your quote : 
> try/catch is for things that go wrong that are outside of your control and not 
> in the normal program flow. 
We can argue that the case we are talking about is not normal program flow.

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: For me preconditions should be the first-of-all process done by one method.

In our case, we can do this for some checks (mailbox name), 
but then we came at a point where we need some data from the current persisted state.

The we need a process before the remaining preconditions, and that's why it's not precondition and Exceptions are here for, IMO.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 1 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: Ok, this exception has been renamed "SystemMailboxNotUpdatableException".

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 2 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: It will also generate the same exception so the description message has been changed in : "Cannot update a system mailbox."

**Comment by mbaechler** [comment]
:+1:

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 3 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: I think the question remains if the system mailbox has not been already created.

**Comment by chibenwa** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Comment by rouazana** [comment]
:+1:

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
:+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 5]
```diff
@@ -130,11 +130,18 @@ private Mailbox getMailbox(String mailboxId, MailboxSession mailboxSession) thro
 
     private void validateMailboxName(MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) throws MailboxNameException {
         char pathDelimiter = mailboxSession.getPathDelimiter();
+        
         if (updateRequest.getName()
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: extract this condition to a method `nameContainsPathDelimiter()`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 6]
```diff
@@ -130,11 +130,18 @@ private Mailbox getMailbox(String mailboxId, MailboxSession mailboxSession) thro
...
                 .filter(name -> name.contains(String.valueOf(pathDelimiter)))
                 .isPresent()) {
             throw new MailboxNameException(String.format("The mailbox '%s' contains an illegal character: '%c'", updateRequest.getName().get(), pathDelimiter));
         }
+        if (updateRequest.getName()
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: extract this condition to `nameMatchSystemMailbox()`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 5 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: done

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 6 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: done

**Comment by chibenwa** [comment]
:+1:

**Comment by mbaechler** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:
